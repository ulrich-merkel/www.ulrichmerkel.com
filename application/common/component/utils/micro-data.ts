/**
 * Es6 module for creating mirco data attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type ItemTypeAttributes = {
    itemScope: boolean;
    itemType: string;
};

/**
 * Get component properties for itemProp and itemType.
 *
 * @param {boolean} itemType - The itemType property if available
 * @returns {object|null} The component props for spreading
 */
export function getItemTypeAttributes(
    itemType?: string
): ItemTypeAttributes | null {
    return isValidString(itemType)
        ? {
              itemScope: true,
              itemType
          }
        : null;
}
