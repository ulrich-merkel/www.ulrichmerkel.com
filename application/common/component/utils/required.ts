/**
 * Es6 module for creating form input required attributes.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isBoolean } from 'lodash';

type RequiredAttributes = {
    'aria-required': boolean;
    required: boolean;
};

/**
 * Get component properties for required form input elements.
 *
 * @param {boolean} required - Whether the input element is required or not
 * @returns {object|null} The component props for spreading
 */
export function getRequiredAttributes(
    required?: boolean
): RequiredAttributes | null {
    return isBoolean(required) && required
        ? {
              'aria-required': true,
              required: true
          }
        : null;
}
