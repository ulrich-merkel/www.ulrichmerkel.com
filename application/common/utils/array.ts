/**
 * Es6 module for handling array related utilities.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */

/**
 * Check if given "input" is a valid array with at least one entry.
 *
 * @param {Array<*>} array - The array to be checked
 * @returns {boolean} Whether this is an array with some entries
 */
export function isValidArray(array?: any[]): boolean {
    return Array.isArray(array) && array.length > 0;
}
