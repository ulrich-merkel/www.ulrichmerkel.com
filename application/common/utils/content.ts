/**
 * Es6 module for helper component.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { get, isArray, isObject, isString } from 'lodash';

import { AVAILABLE_LOCALES } from '../state/intl/duck';

/**
 * Walk through array
 *
 * @private
 * @param {Array} source - The source content
 * @param {Function} callback - The ready callback
 * @returns {Array}
 */
function traverseArray(source, callback) {
    return source.map((value) => {
        return traverse(value, callback); // eslint-disable-line no-use-before-define
    });
}

/**
 * Walk through object
 *
 * @private
 * @param {object} source - The source content
 * @param {Function} callback - The ready callback
 * @returns {object}
 */
function traverseObject(source, callback) {
    const result = {};
    Object.keys(source).forEach((key) => {
        result[key] = traverse(source[key], callback); // eslint-disable-line no-use-before-define
    });
    return result;
}

/**
 * Walk through config
 *
 * @TODO Use functional style, immutable
 *
 * @private
 * @param {object|Array|string} source - The source content
 * @param {Function} callback - The ready callback
 * @returns {*}
 */
function traverse(source, callback) {
    if (isArray(source)) {
        return traverseArray(source, callback);
    }
    if (isObject(source)) {
        return traverseObject(source, callback);
    }
    if (isString(source)) {
        return callback.apply(this, [source]);
    }
    return source;
}

/**
 * @TODO Use functional style, immutable
 *
 * @private
 * @param {object} configContent - The complete content config
 * @param {string} configTranslation - The complete content translation
 * @returns {object}
 */
function translateContent(configContent, configTranslation) {
    if (!configTranslation) {
        return configContent;
    }

    return traverse(configContent, (value) => {
        const translatedString = configTranslation[value];
        if (translatedString) {
            return translatedString;
        }
        return value;
    });
}

/**
 * @param {object} config - The translated config
 * @returns {Function}
 */
function getContentSection(config) {
    return function getContentSectionByKey(key) {
        return get(config, key);
    };
}

/**
 * @TODO use memoize https://addyosmani.com/blog/faster-javascript-memoization/
 *
 * @param {string} [locale=defaultLocale] - The current locale
 * @param {object} config - The translation config
 * @param {string} configKey - The config key
 * @returns {object}
 */
function getTranslatedContent(
    locale = AVAILABLE_LOCALES[0],
    config,
    configKey
) {
    const configContent = get(config, 'content.data');
    const configContentByKey = get(configContent, configKey, configContent);
    const configTranslation = get(config, `${locale.toLowerCase()}.data`);

    return translateContent(configContentByKey, configTranslation);
}

export { getContentSection, getTranslatedContent };
