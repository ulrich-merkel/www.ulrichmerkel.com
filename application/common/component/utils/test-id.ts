/**
 * Es6 module for creating test id related attributes
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { isValidString } from '../../utils/string';

type TestIdAttributes = {
    'data-testid': string;
};

/**
 * Get component properties for testing.
 *
 * @param {string} testId - The test id to be added
 * @returns {object|null} The component props for spreading
 */
export function getTestIdAttributes(testId?: string): TestIdAttributes | null {
    return isValidString(testId)
        ? {
              'data-testid': testId
          }
        : null;
}
