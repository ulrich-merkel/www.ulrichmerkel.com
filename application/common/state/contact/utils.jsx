/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2016
 * @version 0.0.1
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 *
 * @TODO Should be refactored and moved to actions
 *
 * @requires common/utils/validation
 *
 * @changelog
 * - 0.0.1 Basic functions and structure
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';

/**
 * Validation helper to check all inputs.
 *
 * @function
 * @param {object} [state={}] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
function validate(state = {}) {
    return {
        name: required(state.name),
        email: required(state.email) && email(state.email),
        subject: required(state.subject),
        message: required(state.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @function
 * @param {object} state - The current redux contact state
 * @returns {boolean} The validation status
 */
function isValid(state) {
    const validated = validate(state);

    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

function canSendForm(state) {
    const isStateValid = isValid(state);

    return isStateValid
        && state.pristine
        && !state.pending
        && isBrowser(state);
}

export {
    canSendForm,
    isValid,
    validate
};
