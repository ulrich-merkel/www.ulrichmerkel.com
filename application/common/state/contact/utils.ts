/**
 * Es6 module for Redux Architecture.
 * Simple helper component to add utility functions.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 *
 * @see {@link http://redux.js.org/docs/recipes/reducers/ImmutableUpdatePatterns.html}
 */
import { required, email } from '../../utils/validation';
import { isBrowser } from '../../utils/environment';
import { ContactFormType, ValidateType } from './types';

/**
 * Validation helper to check all inputs.
 *
 * @param {object} [contactForm] - The current redux contact state if given
 * @returns {object} The validation status for all inputs as object
 */
export function validate(contactForm: ContactFormType): ValidateType {
    return {
        name: required(contactForm?.name),
        email: required(contactForm?.email) && email(contactForm?.email),
        subject: required(contactForm?.subject),
        message: required(contactForm?.message)
    };
}

/**
 * Validation single or all input element(s), depending on parameter input.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} The validation status
 */
export function isValid(contactForm: ContactFormType): boolean {
    const validated = validate(contactForm);
console.log({validated, contactForm})
    return !Object.keys(validated).some(function someKey(key) {
        return validated[key] !== undefined && !validated[key];
    });
}

/**
 * Check if the form can be send.
 *
 * @param {object} [contactForm] - The current redux contact state
 * @returns {boolean} Whether the submit button is enabled or not
 */
export function canSendForm(contactForm: ContactFormType): boolean {
    const isStateValid = isValid(contactForm);

    return (
        isStateValid &&
        contactForm.pristine &&
        !contactForm.pending &&
        isBrowser(contactForm)
    );
}
