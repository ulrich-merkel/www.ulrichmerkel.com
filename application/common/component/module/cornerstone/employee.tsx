/**
 * Es6 module for a cornerstone employee block.
 *
 * @file
 * @module
 *
 * @author hello@ulrichmerkel.com (Ulrich Merkel), 2021
 */
import { default as React, FunctionComponent } from 'react';
import shortid from 'shortid';

import { isValidArray } from '../../../utils/array';
import { ModuleCornerstoneItemEmployee } from './item-employee';

type Props = {
    list?: {
        description?: string[];
        headline?: string;
        lead?: string;
        offset?: string;
        timeEnd?: string;
        timeStart?: string;
    }[];
};

/**
 * Function representing a cornerstone employee block.
 *
 * @function
 * @param {object} [props] - The current component props
 * @returns {Array<ReactElement>} React component markup
 */
export const ModuleCornerstoneEmployee: FunctionComponent<Props> = (props) => {
    const { list } = props;

    if (!isValidArray(list)) {
        return null;
    }

    return list.map(function fnMap(value, index) {
        const {
            description,
            headline,
            lead,
            offset,
            timeEnd,
            timeStart
        } = value;

        return (
            <ModuleCornerstoneItemEmployee
                key={shortid.generate()}
                cssModifier={index % 2 === 0 ? 'left' : 'right'}
                {...{
                    description,
                    headline,
                    lead,
                    offset,
                    timeEnd,
                    timeStart
                }}
            />
        );
    });
};
