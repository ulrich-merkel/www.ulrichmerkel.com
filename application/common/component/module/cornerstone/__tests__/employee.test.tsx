import * as React from 'react';
import { render } from '../../../../../__tests__/utils/test-utils';
import { ModuleCornerstoneEmployee } from '../employee';

describe('ModuleCornerstoneEmployee', function fnDescribe() {
    const props = {
        list: [
            {
                headline: 'Headline item employee',
                lead: 'Lead  item employee',
                timeStart: '20161212',
                timeEnd: '20161224',
                description: [
                    'Description item employee 1',
                    'Description item employee 2'
                ]
            },
            {
                headline: 'Headline item employee',
                lead: 'Lead  item employee',
                timeStart: '20161212',
                timeEnd: '20161224',
                description: [
                    'Description item employee 1',
                    'Description item employee 2'
                ]
            },
            {
                headline: 'Headline item employee',
                lead: 'Lead  item employee',
                timeStart: '20161212',
                timeEnd: '20161224',
                description: [
                    'Description item employee 1',
                    'Description item employee 2'
                ]
            }
        ]
    };

    it('should render correctly', function fnIt() {
        const { asFragment } = render(<ModuleCornerstoneEmployee {...props} />);
        expect(asFragment()).toMatchSnapshot();
    });
    it('should render correctly with defaults', function fnIt() {
        const { asFragment } = render(<ModuleCornerstoneEmployee />);
        expect(asFragment()).toMatchSnapshot();
    });
});
