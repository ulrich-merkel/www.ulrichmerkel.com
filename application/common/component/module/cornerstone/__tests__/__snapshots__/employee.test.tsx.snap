// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleCornerstoneEmployee should render correctly 1`] = `
<DocumentFragment>
  <li
    class="c-list-item m-cornerstone__item--left"
    itemprop="itemListElement"
    itemscope=""
    itemtype="https://schema.org/EmployeeRole"
  >
    <div
      class="m-cornerstone__description"
    >
      <div
        class="m-cornerstone__description-content"
      >
        <h4
          class="is-centered c-type--h4 m-cornerstone__headline"
          itemprop="roleName"
        >
          Headline item employee
        </h4>
        <p
          class="c-type--p m-cornerstone__company"
        >
          <strong>
            Lead  item employee
          </strong>
        </p>
        <p
          class="c-type--p m-cornerstone__time"
          itemprop="description"
        >
          (
          <time
            class="c-time"
            itemprop="startDate"
          >
            20161212
          </time>
           -
          <time
            class="c-time"
            itemprop="endDate"
          >
            20161224
          </time>
          )
        </p>
        <p
          class="c-type--p m-cornerstone__text"
          itemprop="description"
        >
          Description item employee 1
        </p>
        <p
          class="c-type--p m-cornerstone__text"
          itemprop="description"
        >
          Description item employee 2
        </p>
      </div>
    </div>
    <div
      class="m-cornerstone__bubble"
    />
  </li>
  <li
    class="c-list-item m-cornerstone__item--right"
    itemprop="itemListElement"
    itemscope=""
    itemtype="https://schema.org/EmployeeRole"
  >
    <div
      class="m-cornerstone__description"
    >
      <div
        class="m-cornerstone__description-content"
      >
        <h4
          class="is-centered c-type--h4 m-cornerstone__headline"
          itemprop="roleName"
        >
          Headline item employee
        </h4>
        <p
          class="c-type--p m-cornerstone__company"
        >
          <strong>
            Lead  item employee
          </strong>
        </p>
        <p
          class="c-type--p m-cornerstone__time"
          itemprop="description"
        >
          (
          <time
            class="c-time"
            itemprop="startDate"
          >
            20161212
          </time>
           -
          <time
            class="c-time"
            itemprop="endDate"
          >
            20161224
          </time>
          )
        </p>
        <p
          class="c-type--p m-cornerstone__text"
          itemprop="description"
        >
          Description item employee 1
        </p>
        <p
          class="c-type--p m-cornerstone__text"
          itemprop="description"
        >
          Description item employee 2
        </p>
      </div>
    </div>
    <div
      class="m-cornerstone__bubble"
    />
  </li>
  <li
    class="c-list-item m-cornerstone__item--left"
    itemprop="itemListElement"
    itemscope=""
    itemtype="https://schema.org/EmployeeRole"
  >
    <div
      class="m-cornerstone__description"
    >
      <div
        class="m-cornerstone__description-content"
      >
        <h4
          class="is-centered c-type--h4 m-cornerstone__headline"
          itemprop="roleName"
        >
          Headline item employee
        </h4>
        <p
          class="c-type--p m-cornerstone__company"
        >
          <strong>
            Lead  item employee
          </strong>
        </p>
        <p
          class="c-type--p m-cornerstone__time"
          itemprop="description"
        >
          (
          <time
            class="c-time"
            itemprop="startDate"
          >
            20161212
          </time>
           -
          <time
            class="c-time"
            itemprop="endDate"
          >
            20161224
          </time>
          )
        </p>
        <p
          class="c-type--p m-cornerstone__text"
          itemprop="description"
        >
          Description item employee 1
        </p>
        <p
          class="c-type--p m-cornerstone__text"
          itemprop="description"
        >
          Description item employee 2
        </p>
      </div>
    </div>
    <div
      class="m-cornerstone__bubble"
    />
  </li>
</DocumentFragment>
`;

exports[`ModuleCornerstoneEmployee should render correctly with defaults 1`] = `<DocumentFragment />`;
